Heterogeneous Runtimes General Open Problem (#10)
protocol/ResNetLab: OPEN_PROBLEMS/HETEROGENEOUS_RUNTIMES.md

> # Heterogeneous Runtimes

Reviewer: ```suggestion
# Networking in Heterogeneous Runtimes
```

## Short Description

> Annual global IP traffic will reach 396 exabytes per month. The number of devices connected to the internet is expected to double in the same time. Access to high speed broadband connectivity is limited by poor last mile connections. Optic-fiber deployments are cost-prohibitive and slow to deploy.

Reviewer: ```suggestion
Annual global IP traffic will reach 396 exabytes per month. The number of devices connected to the internet is expected to double in the same time. Access to high speed broadband connectivity is limited by poor last mile connections. Optic-fiber deployments are cost-prohibitive and slow to deploy [1](please provide reference).
```

Edge  computing  has  emerged  as  a  distributed  computing paradigm to overcome practical scalability limits of cloud computing.  The main principle of edge computing is to leverage computational resources outside of the cloud to perform computations closer to data sources, avoiding unnecessary data transfers to the cloud and enabling faster responses for clients. Given the enormous amount of data that is expected to be produced at the edge of the network (by end-user devices), the edge-computing principle builds on the fact that “it is cheaper to bring computation to data, rather than data to computation”.

Data storage and computation points at the edge of the network, be it in cell stations, WiFi access points, or even the same user devices that produce the data ultimately form a peer-to-peer network.

> *Both IPFS and libp2p can have a pivotal role in this new trend of distributed technologies and expansion of the Internet beyond traditional boundaries*. Both technologies already have implementations in several programming languages and are compatible for their execution in a great gamut of runtimes (desktop, browser, cloud, etc.) and network conditions (public IP addresses, behind NATs, etc.). Despite this success, there is still a lot of work to be done to make IPFS and libp2p compatible with any runtime and network condition in a way that offers a flawless experience, just like one expects when using the net or http package of any language. This open problem is built upon *the vision of making libp2p / IPFS modules and protocols the de-facto distributed substrate for connected devices in the near-future Internet*.

Reviewer: ```suggestion
*Both IPFS and libp2p can have a pivotal role in this new trend of distributed technologies and expansion of the Internet beyond traditional boundaries*. Both technologies already have implementations in several programming languages and are compatible for their execution in a great gamut of runtimes (desktop, browser, cloud, etc.) and network conditions (public IP addresses, behind NATs, etc.). Nevertheless, there is significant work to be done to make IPFS and libp2p compatible with every runtime and network condition in a way that offers a flawless experience, just like one expects when using the net or http package of any language. This open problem is built upon *the vision of making libp2p / IPFS modules and protocols the de-facto distributed substrate for connected devices in the near-future Internet*.
```

At high-level, the aim of this research endeavor is to:

> -   Enable the execution of libp2p nodes and its underlying protocols anywhere (any device and any runtime).
>     - [Are we yet?](https://wiki.mozilla.org/Areweyet)

Reviewer: ```suggestion

-  Enable libp2p to leverage the available protocols from the multiple runtimes, so that it can execute and form a p2p network. This will require a coordinated approach to cover as much ground possible (possible approach inspired by [Are we yet?](https://wiki.mozilla.org/Areweyet) initiatives)
```

> -   Allow global connectivity between devices. Any two libp2p nodes should be able to communicate using a compatible network link available without having to rely on a stable Internet connection (direct connectivity, wireless transports, mesh connectivity, etc.). All protocol implementations should be compatible.

Reviewer: ```suggestion
-   Allow global connectivity between devices. Any two libp2p nodes should be able to communicate using a compatible network link available without having to rely on a connection to the broad Internet (e.g direct connectivity through wireless transports, mesh connectivity, etc.). 
```

> -   Libp2p and IPFS should be "offline-first" and enable "offline-first" applications. Temporary loss of  Internet connection shouldn't break their operation, and they should be able to seamlessly recover from unstable, intermittent network conditions (in many scenarios this is already the case).

Reviewer: ```suggestion
```

This is already the case, unless there is a litmus test I'm missing

-   (Bonus) Libp2p/IPFS should include a general execution framework to be easily extensible for any runtime using the same format.

> -   [Extensibility through Web Assembly](https://istio.io/latest/blog/2020/wasm-announce/)
> -   [Web Assembly for proxies (ABI specification)](https://github.com/proxy-wasm/spec)
> -   [Self-described blocks (embedded IPLD codecs)](https://hackmd.io/AAfd9WnWQZSC7HT7Pr5G9A?view)

Reviewer: ```suggestion
  -   [Extensibility through Web Assembly](https://istio.io/latest/blog/2020/wasm-announce/)
  -   [Web Assembly for proxies (ABI specification)](https://github.com/proxy-wasm/spec)
  -   [Self-described blocks (embedded IPLD codecs)](https://hackmd.io/AAfd9WnWQZSC7HT7Pr5G9A?view)
```

## Long Description
This open problem can be divided in the following subproblems:

> ### 📲️ Area II: Runtimes

Reviewer: ```suggestion
### 📲️ Area I: Runtimes
```

By the end of this research effort we should consider at least supporting the following runtime categories (each category includes a list of related projects):
-   Browsers and Desktop
-   Embedded systems (IoT, low powered devices)
    -   [IPFS Tiny](https://gitlab.com/librespacefoundation/ipfs-tiny/-/wikis/home)
    -   [Embed-ipfs](https://github.com/ipfs-rust/ipfs-embed)
-   Routing devices:
    -   [OpenWrt](https://openwrt.org/)
    -   [Gotenna](https://gotenna.com/)
    -   [Liberouter](http://www.liberouter.mobi/)
-   Mobile devices:
    -   [Thali Project](http://thaliproject.org/)
-   VR/AR devices.
-   Trusted Execution Environments
    -   [Enarx](https://github.com/enarx/enarx/wiki/Enarx-Introduction)
    -   [T-Rust](http://t-rust.com/#/): [TEA (Trusted Execution and Attestation)](http://t-rust.org/#/doc_list/What_is_TEA%3F%2FREADME.md)

We may be able to support many of the aforementioned with the same protocol implementations (e.g. routing devices and embedded systems). This work will set a strong foundation for the rest of the projects to come.

#### What defines a complete solution?

> -   XXX

Reviewer: ```suggestion
- TBD
```

#### References: 
-   Improve offline support: <https://github.com/libp2p/notes/issues/16>
-   Offline message queue: <https://github.com/libp2p/notes/issues/2>
-   IPFS CRDT datastore: <https://github.com/ipfs/go-ds-crdt>
-   Optimize Circuit Relays to make it packet-oriented in libp2p: <https://github.com/libp2p/go-libp2p/issues/1019>
-   NAT Traversal effors around libp2p: <https://github.com/libp2p/go-libp2p/issues?q=is%3Aissue+is%3Aopen+NAT+Traversal>

### 🛠️ (Bonus) Area V: Extensibility / P2P VM a.k.a InterPlanetary Runtime
This area of research is still a work in progress, and it may becoming a full-fledge open problem of its own, but for now we still consider it as an area inside the heterogeneous runtime problem.
Maintaining and growing libp2p protocols for the gamut of runtimes and connections we want to support can be cumbersome. This is why we should consider embedding a common execution framework to every libp2p implementation so libp2p modules are compiled targeting this common framework and can be thus used in any of the runtimes. With this, protocols and modules are implemented once and used anywhere.

> Not only protocol implementations can benefit from embedding an *"InterPlanetary Runtime* in libp2p nodes, but also any other standard computation in the ecosystem such as IPLD codecs. IPLD codecs can be implemented targeting the *IP Runtime* so they don't have to also be implemented for every libp2p implementation and runtime. This opens the door to future lines of research such as *self-describing blocks*, where the block descriptions embeds the type of data and a pointer to the codec (which can be another block in stored in the network); *content-addressable code executions*, where code snippets are stored in the network, and data is specified in the code as pointers to blocks in th network; *computation delegation*, as once every node in the network shares the same common runtime and has a way to retrieve code and data directly from the network with a standard interface, delegating the execution of code to other's is as easy as sharing a link to the code and the data.

Reviewer: This could likely go into the IPLD Open Problem (which hasn't been written yet). It can help on the interoperability, but seems a non priority for this Open Problem

#### What defines a complete solution?
-   TBD

### References
- Wasmtime (Wasm runtime): <https://wasmtime.dev/>
- UnisonWeb: Example of content-addressable programming language: <https://www.unisonweb.org/>
- Wasmer (Wasm runtime): <https://wasmer.io/>
- WASM ❤️ IPLD: <https://hackmd.io/AAfd9WnWQZSC7HT7Pr5G9A?edit>
- For this project we can take inspiration from the[Wasm extensibility approach](https://istio.io/latest/blog/2020/wasm-announce/) currently introduced in Istio and Envoy. Thus, our libp2p node implementations would expose the basic operation with a common runtime framework from which new Wasm modules are instantiated and run to extend the operation of our node with additional modules and protocols.
- Wasm bindgen designs: <https://rustwasm.github.io/docs/wasm-bindgen/contributing/design/index.html>

> ### 🛰️ (Bonus) Area VI: Space

Reviewer: ```suggestion
### 🛰️ (Bonus) Area VI: High Latency Connection (e.g. Space)
```
